library/std/src/io/buffered/bufreader.rs

```diff
@@ -308,6 +308,14 @@ impl<R: Read> Read for BufReader<R> {
     unsafe fn initializer(&self) -> Initializer {
         self.inner.initializer()
     }
+
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        self.inner.read_to_end(buf)
+    }
+
+    fn read_to_string(&mut self, buf: &mut String) -> io::Result<usize> {
+        self.inner.read_to_string(buf)
+    }
```

Comment:
These functions should take the internal buffer into account and use its data before deferring to the inner reader.

Reply:
Thanks, that's a good catch. I've fixed them and added corresponding unit tests to verify the fixes. Could you take a look at what I did? Getting these right was tricky. I'm wondering if I missed a simpler way.

Reply:
I don't love how `BufReader::read_to_string` reads into a side byte buffer, checks if it's valid UTF-8, and then appends the bytes to the passed in `String` if the UTF-8 validation succeeds. I wish I could do what [`append_to_string`](https://doc.rust-lang.org/src/std/io/mod.rs.html#335) does and append bytes directly to the input `String`, but I don't think it's safe to do so. `append_to_string` uses unsafe code and is only safe because it calls a trusted version of `read_to_end` that it knows will only append data. I can't do the same thing in `BufReader` because I don't trust `self.inner.read_to_end`, so there's an extra allocation/memcpy.

Reply:
I think you could `append_to_string` the internal buffer, where you know you'll only append to the string, and then call `self.inner.read_to_string` to read what is left of the reader.
In this case, you don't have to trust `self.inner.read_to_end`.